presto-hive/src/main/java/com/facebook/presto/hive/metastore/RecordingHiveMetastore.java

```diff
@@ -0,0 +1,582 @@
+import com.facebook.presto.spi.statistics.ColumnStatisticType;
+import com.facebook.presto.spi.type.Type;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.json.ObjectMapperProvider;
+import org.weakref.jmx.Managed;
+
+import javax.annotation.concurrent.Immutable;
+import javax.inject.Inject;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.hive.metastore.HivePartitionName.partition;
+import static com.facebook.presto.hive.metastore.HiveTableName.table;
+import static com.facebook.presto.hive.metastore.PartitionFilter.partitionFilter;
+import static com.facebook.presto.spi.StandardErrorCode.CONFIGURATION_INVALID;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class RecordingHiveMetastore
+        implements ExtendedHiveMetastore
+{
+    private final ExtendedHiveMetastore delegate;
+    private final String recordingPath;
+    private final boolean replay;
+
+    private Optional<List<String>> allDatabases;
+
+    private final Cache<String, Optional<Database>> databaseCache;
+    private final Cache<HiveTableName, Optional<Table>> tableCache;
+    private final Cache<HiveTableName, PartitionStatistics> tableStatisticsCache;
+    private final Cache<Set<HivePartitionName>, Map<String, PartitionStatistics>> partitionStatisticsCache;
+    private final Cache<String, Optional<List<String>>> allTablesCache;
+    private final Cache<String, Optional<List<String>>> allViewsCache;
+    private final Cache<HivePartitionName, Optional<Partition>> partitionCache;
+    private final Cache<HiveTableName, Optional<List<String>>> partitionNamesCache;
+    private final Cache<PartitionFilter, Optional<List<String>>> partitionNamesByPartsCache;
+    private final Cache<Set<HivePartitionName>, Map<String, Optional<Partition>>> partitionsByNamesCache;
+    private final Cache<String, Set<String>> rolesCache;
+    private final Cache<UserDatabaseKey, Set<HivePrivilegeInfo>> databasePrivilegesCache;
+    private final Cache<UserTableKey, Set<HivePrivilegeInfo>> tablePrivilegesCache;
+
+    @Immutable
+    public static class Pair<K, V>
+    {
+        final K key;
+        final V value;
+
+        @JsonCreator
+        public Pair(@JsonProperty("key") K key, @JsonProperty("value") V value)
+        {
+            this.key = key;
+            this.value = value;
```

Comment:
can any of these be null?

Reply:
I'm not sure. Let's not assume that for now

Reply:
sure. Please add `rnn` checks here